### PR TITLE
[mlir][vector] Refine vector.transfer_read hoisting/forwarding

### PR DESCRIPTION
**Pull request by banach-space**
Make sure that when analysing a `vector.transfer_read` that's a candidate for either hoisting or store-to-load forwarding, `memref.collapse_shape` Ops are correctly included in the alias analysis. This is done by either
* making sure that relevant users are taken into account, or
* source Ops are correctly identified.

### CODE REVIEW

File: mlir/test/Dialect/Vector/vector-transferop-opt.mlir [hunk 1]
```diff
@@ -215,4 +215,41 @@ func.func @forward_dead_store_negative(%arg0: i1, %arg1 : memref<4x4xf32>,
...
+}
+
+
+// Regression test - the following _potential forwarding_ of %1 to the final
+// `vector.transfer_write` would not be safe:
+//         %1 = vector.transfer_read %subview
+//         vector.transfer_write %1, %alloca
+//         vector.transfer_write %vec, %collapse_shape 
+//         vector.transfer_write %1, %subview
+// Indeed, %alloca and %collapse_shape alias and hence %2 != %1.
```
Comment (matthias-springer) [inline mlir/test/Dialect/Vector/vector-transferop-opt.mlir:227]: There is no `%2` in this example.

**Review by matthias-springer — APPROVED** [review]
This fixes hoisting for `collapse_shape`, but it is still broken for other ops that introduce aliasing (e.g., `expand_shape`, `arith.select`, ...). We should query an alias analysis in `noAliasingUseInLoop` instead of hard-coding ops. Still, better than before so I'd say let's land it if it's blocking.

File: mlir/test/Dialect/Vector/vector-transferop-opt.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transferop-opt.mlir]: Thanks, clarified in the latest fixup.

**Comment by banach-space** [comment]
> This fixes hoisting for `collapse_shape`, but it is still broken for other ops that introduce aliasing (e.g., `expand_shape`, `arith.select`, ...). We should query an alias analysis in `noAliasingUseInLoop` instead of hard-coding ops. Still, better than before so I'd say let's land it if it's blocking.

Thanks! Agreed that a more generic would be better, but right now this is causing our convs to be broken - I really appreciate you being flexible here 🙏🏻 .

**Comment by matthias-springer** [comment]
Btw, can you perform the hoisting on tensors instead of memrefs? There are no aliasing issues when hoisting before bufferization. I'm trying to see if it's worth generalizing the hoisting infrastructure, making it interface-based, and moving it to `mlir/Transforms`.

**Comment by banach-space** [comment]
> Btw, can you perform the hoisting on tensors instead of memrefs? There are no aliasing issues when hoisting before bufferization. 

Tl;Dr Possibly. It looks like in many cases hoisting happens _before_ bufferization, but not always. 

I'm experimenting with some optimisations for Linalg Convs and use IREE to drive this. Looking at 
* [Codegen/LLVMCPU/Passes.cpp](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp) (I only use CPU pipelines), 

[addBufferizePasses](https://github.com/openxla/iree/blob/3343f24363dffa7e4602305d33627302d8c0bd6c/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L128-L134) is usually run after e.g. [createHoistRedundantVectorTransfersPass](https://github.com/openxla/iree/blob/3343f24363dffa7e4602305d33627302d8c0bd6c/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L569-L570). However, in the case of convolutions, [OptimizeVectorTransferPass](https://github.com/openxla/iree/blob/3343f24363dffa7e4602305d33627302d8c0bd6c/compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp#L139-L142) is [created](https://github.com/openxla/iree/blob/3343f24363dffa7e4602305d33627302d8c0bd6c/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L579-L583) after bufferization and that pass also [includes](https://github.com/openxla/iree/blob/3343f24363dffa7e4602305d33627302d8c0bd6c/compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp#L83) `hoistRedundantVectorTransfers`.

I've only tried few pipelines so far, so might be missing something. Also, I have a few experiments "downstream" on top of this that trigger these aliasing issue. So, on hand, having this fix in-tree helps me focus on other issues, but I might be doing something daft myself. 

I'll add a few TODOs before landing this. Mostly as a reminder that a more generic approach should be implemented instead.
